recipes/fmt/all/conanfile.py

```diff
@@ -64,65 +58,63 @@ def config_options(self):
             self.options.with_os_api = False
 
     def configure(self):
-        if self.options.header_only:
-            del self.options.fPIC
-            del self.options.shared
-            del self.options.with_os_api
-        elif self.options.shared:
-            del self.options.fPIC
-
-    def validate(self):
-        if self.options.get_safe("shared") and self._is_msvc and "MT" in msvc_runtime_flag(self):
-            raise ConanInvalidConfiguration(
-                "Visual Studio build for shared library with MT runtime is not supported"
-            )
+        try:
+            if self.options.header_only:
+                del self.options.fPIC
+                del self.options.shared
+                del self.options.with_os_api
+            elif self.options.shared:
+                del self.options.fPIC
+        except Exception:
+            pass
 
     def package_id(self):
-        if self.options.header_only:
+        if self.info.options.header_only:
             self.info.header_only()
         else:
             del self.info.options.with_fmt_alias
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
-
-    def _configure_cmake(self):
-        if self._cmake:
-            return self._cmake
-        self._cmake = CMake(self)
-        self._cmake.definitions["FMT_DOC"] = False
-        self._cmake.definitions["FMT_TEST"] = False
-        self._cmake.definitions["FMT_INSTALL"] = True
-        self._cmake.definitions["FMT_LIB_DIR"] = "lib"
-        if self._has_with_os_api_option:
-            self._cmake.definitions["FMT_OS"] = self.options.with_os_api
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+        get(self, **self.conan_data["sources"][str(self.version)], destination=self.source_folder, strip_root=True)
 
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        apply_conandata_patches(self)
         if not self.options.header_only:
-            cmake = self._configure_cmake()
+            cmake = CMake(self)
+            # FIXME : https://github.com/conan-io/conan/issues/11476
+            cache_entries = {
+                "FMT_DOC": "False",
+                "FMT_TEST": "False",
+                "FMT_INSTALL": "True",
+                "FMT_LIB_DIR": "lib"
+            }
+            if self._has_with_os_api_option:
+                cache_entries["FMT_OS"] = self.options.with_os_api
```

Comment:
This should be at the ``CMakeToolchain`` level, that allows local user builds to be consistent with options

Reply:
I think this can't be changed until c3i uses Conan 1.50 and then must be translated to https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#cache-variables

Reply:
CCI uses 1.49 ATM

Reply:
there was a bug using `tc.variables` had a zero effect. it causes CI to build more things than needed (e.g. it build tests that were intended to be disabled via `FMT_TEST`).
it turns out to be that diff:
https://github.com/fmtlib/fmt/blob/f94b7364b9409f05207c3af3fa4666730e11a854/CMakeLists.txt#L41-L49
https://github.com/fmtlib/fmt/blob/b6f4ceaed0a0a24ccf575fab6c56dd50ccf6f1a9/CMakeLists.txt#L66-L81
in general it harmless in this particular case (just causes longer builds), but in other cases it can produce a wrong package if some option is ignored.
more info at https://github.com/conan-io/conan/issues/11476

Reply:
Ahhh, there will need to be a hook so people dont use the other one in CCI 🙈 

The options before project is an obvious limitation + it's easy to imagine the separate steps for the c3i could be problematic

Thanks for filling in the details